packages/order-utils/src/asset_proxy_utils.ts

```diff
@@ -100,50 +114,67 @@ export const assetProxyUtils = {
         const tokenAddress = assetProxyUtils.decodeAddress(encodedTokenAddress);
-        const tokenIdOffset = ERC721_PROXY_METADATA_BYTE_LENGTH - 1;
-        const encodedTokenId = encodedProxyMetadata.slice(addressOffset, tokenIdOffset);
+        const encodedTokenId = encodedAssetData.slice(tokenIdOffset, receiverDataLengthOffset);
         const tokenId = assetProxyUtils.decodeUint256(encodedTokenId);
-        const erc721ProxyData = {
+        const nullData = '0x';
+        let receiverData = nullData;
+        const lengthUpToReceiverDataLength = receiverDataLengthOffset + 1;
+        if (encodedAssetData.byteLength > lengthUpToReceiverDataLength) {
+            const encodedReceiverDataLength = encodedAssetData.slice(receiverDataLengthOffset, receiverDataOffset);
+            const receiverDataLength = assetProxyUtils.decodeUint256(encodedReceiverDataLength);
+            const lengthUpToReceiverData = receiverDataOffset + 1;
+            const expectedReceiverDataLength = new BigNumber(encodedAssetData.byteLength - lengthUpToReceiverData);
+            if (!receiverDataLength.equals(expectedReceiverDataLength)) {
+                throw new Error(
+                    `Data length (${receiverDataLength}) does not match actual length of data (${expectedReceiverDataLength})`,
+                );
+            }
+            const encodedReceiverData = encodedAssetData.slice(
+                receiverDataOffset,
+                receiverDataOffset + receiverDataLength.toNumber(),
+            );
+            receiverData = ethUtil.bufferToHex(encodedReceiverData);
+        }
+        const erc721AssetData: ERC721AssetData = {
             assetProxyId,
             tokenAddress,
             tokenId,
+            receiverData,
         };
-        return erc721ProxyData;
+        return erc721AssetData;
     },
-    decodeProxyDataId(proxyData: string): AssetProxyId {
-        const encodedProxyMetadata = ethUtil.toBuffer(proxyData);
-        if (encodedProxyMetadata.byteLength < 1) {
+    decodeAssetDataId(assetData: string): AssetProxyId {
+        const encodedAssetData = ethUtil.toBuffer(assetData);
+        if (encodedAssetData.byteLength < 1) {
             throw new Error(
                 `Could not decode Proxy Data. Expected length of encoded data to be at least 1. Got ${
-                    encodedProxyMetadata.byteLength
+                    encodedAssetData.byteLength
                 }`,
             );
         }
-        const encodedAssetProxyId = encodedProxyMetadata.slice(-1);
+        const encodedAssetProxyId = encodedAssetData.slice(-1);
         const assetProxyId = assetProxyUtils.decodeAssetProxyId(encodedAssetProxyId);
         return assetProxyId;
     },
-    decodeProxyData(proxyData: string): ProxyData {
-        const assetProxyId = assetProxyUtils.decodeProxyDataId(proxyData);
+    decodeAssetData(assetData: string): AssetData {
```

Comment:
I think this could be better with a [Union type](http://www.typescriptlang.org/docs/handbook/advanced-types.html) of ERC20AssetData or ERC721AssetData. This will parse the proxy id out and the token address out, but it doesn't tell me what it is.